rfcs/2023-04-18-bundle-check.md

```diff
@@ -0,0 +1,124 @@
+- Bundlemon allows us to pass a config like this in `package.json`:
+```json
+{
+  "files": [
+    {
+      "path": "./build/components/index.web.js",
+      "maxSize": "200kB"
+    },
+    {
+      "path": "./build/tokens/index.web.js",
+      "maxSize": "100kB"
+    },
+    {
+      "path": "./build/components/index.native.js",
+      "maxSize": "200kB"
+    },
+    {
+      "path": "./build/tokens/index.native.js",
+      "maxSize": "100kB"
+    }
+  ]
+}
+```
+- We will need to add a github action to run bundlemon on every PR. This can be a part of `blade-validate`(for PR) + `release`(for master push) workflow or a separate `bundle-check` workflow(that triggers on PR & master push).
+- This github action will pass or fail based on the limits defined in the config.
+- Bundlemon posts a comment on the PR like this:
+
+  <img src="./images/bundle-size/pr-comment.png" width="800" />
+
+- On visiting their report you can see a detailed view like this:
+  <img src="./images/bundle-size/bundle-details.png" width="800" />
+
+- You can also visualize the bundle history branch-wise like this:
+  <img src="./images/bundle-size/bundle-history.png" width="800" />
+
+### Tree Shaking Check
+We will achieve the following with this:
+
+3. Ensure our tree-shaking works as expected by adding a check to ensure that unused components are not included in the bundle.
+
+- We will use [size-limit](https://github.com/ai/size-limit) for this
+- size-limit is a tool that helps us to keep an eye on our bundle size. It works by comparing the size of our bundles against a predefined limit. If the size of the bundle exceeds the limit, the build fails.
+- size-limit creates an empty project and imports the component that we want. It then creates a bundle and checks the size of the bundle. If the size of the bundle exceeds the limit, the build fails.
+- This will enable us to limit importing only a small set of components like `{ Button, Text }` and check the bundle size to ensure that tree-shaking is working as intended and the overall bundle size is limited to the imported components.
+- size-limit configuration looks like this:
+```json
+  "size-limit": [
+    {
+      "path": "./build/components/index.web.js",
+      "import": "{ Button, Text }",
+      "limit": "100 kB"
+    }
+  ]
+```
+- We are not using size-limit to do a total bundle size check or master diff because it doesn't support it as well as bundlemon does. Bundlemon has an infra set up that stores master bundle size and diffs it on every PR. We can't do that with size-limit.
+
+# Drawbacks/Constraints
+- Bundlemon is a free service but it also seems unmaintained since Jan 2022.
+- If Bundlemon service goes down, we will not be able to get any master diffs but we will still be able to get the total bundle size limit check.
+- Alternate to using bundlemon is setting up our own infra to store master bundle size and diff it on every PR. This would also be a lot of work and we would have to maintain it.
```

Comment:
i think this shouldn't be that difficult. what we essentially need is downloading the npm package ➡️ unzipping it ➡️ and then doing the filesize on it. but we can revisit once we face issues with bundlemon.

Reply:
But yeah, we will revisit this if bundlemon causes any issues for us. For now its giving us the ideal solution with minimal efforts